File: server/src/main/kotlin/com/larsreimann/api_editor/server/validation/AnnotationValidator.java
package com.larsreimann.api_editor.server.validation;

import com.larsreimann.api_editor.server.data.*;

import java.util.*;

public class AnnotationValidator {
    AnnotatedPythonPackage annotatedPythonPackage;

    List<AnnotationError> validationErrors;

    private static final String ANNOTATION_PREFIX = "com.larsreimann.api_editor.server.data.";
    private static final String ANNOTATION_CLASS_SUFFIX = "Annotation";

    private static final Map<String, Set<String>> possibleCombinations;
    private static final Set<String> possibleClassAnnotations = Set.of(
        "Move", "Rename", "Unused"
    );
    private static final Set<String> possibleClassFunctionAnnotations = Set.of(
        "CalledAfter", "Group", "Rename", "Unused"
    );
    private static final Set<String> possibleGlobalFunctionAnnotations = Set.of(
        "CalledAfter", "Group", "Move", "Rename", "Unused"
    );
    private static final Set<String> possibleFunctionParameterAnnotations = Set.of(
        "Boundary", "Constant", "Enum", "Optional", "Rename", "Required"
    );
    private static final Set<String> possibleConstructorParameterAnnotations = Set.of(
        "Attribute", "Boundary", "Constant", "Enum", "Optional", "Rename", "Required"
    );

    static {
        possibleCombinations = new HashMap<>();
        possibleCombinations.put("Attribute", Set.of(
            "Boundary", "Enum", "Rename"
        ));
        possibleCombinations.put("Boundary", Set.of(
            "Attribute", "Group", "Optional", "Rename", "Required"
        ));
        possibleCombinations.put("CalledAfter", Set.of(
            "CalledAfter", "Group", "Move", "Rename"
        ));
        possibleCombinations.put("Constant", Set.of());
        possibleCombinations.put("Enum", Set.of(
            "Attribute", "Group", "Optional", "Rename", "Required"
        ));
        possibleCombinations.put("Group", Set.of(
            "Boundary", "CalledAfter", "Enum", "Group", "Move", "Optional", "Rename", "Required"
        ));
        possibleCombinations.put("Move", Set.of(
            "CalledAfter", "Group", "Rename"
        ));
        possibleCombinations.put("Optional", Set.of(
            "Boundary", "Enum", "Group", "Rename"
        ));
        possibleCombinations.put("Rename", Set.of(
            "Attribute", "Boundary", "CalledAfter", "Enum", "Group", "Move", "Optional", "Required"
        ));
        possibleCombinations.put("Required", Set.of(
            "Boundary", "Enum", "Group", "Rename"
        ));
        possibleCombinations.put("Unused", Set.of());
    }

    public AnnotationValidator(AnnotatedPythonPackage inputPackage) {
        annotatedPythonPackage = inputPackage;
        validationErrors = List.of();
    }

    public List<AnnotationError> getValidationErrors() {
        validatePackage();
        return validationErrors;
    }

    private void validateClassFunction(AnnotatedPythonFunction annotatedPythonFunction) {
        annotatedPythonFunction.getParameters().forEach(
            (parameter) -> validateFunctionParameter(parameter, annotatedPythonFunction.getQualifiedName())
        );
        EditorAnnotation[] functionAnnotations = annotatedPythonFunction.getAnnotations().toArray(new EditorAnnotation[0]);
        validateAnnotationsValidOnTarget(functionAnnotations, AnnotatableDeclaration.CLASS_FUNCTION, annotatedPythonFunction.getQualifiedName());
        validateAnnotationCombinations(functionAnnotations, annotatedPythonFunction.getQualifiedName());
    }

    private void validateGlobalFunction(AnnotatedPythonFunction annotatedPythonFunction) {
        annotatedPythonFunction.getParameters().forEach(
            (parameter) -> validateFunctionParameter(parameter, annotatedPythonFunction.getQualifiedName())
        );
        EditorAnnotation[] functionAnnotations = annotatedPythonFunction.getAnnotations().toArray(new EditorAnnotation[0]);
        validateAnnotationsValidOnTarget(functionAnnotations, AnnotatableDeclaration.GLOBAL_FUNCTION, annotatedPythonFunction.getQualifiedName());
        validateAnnotationCombinations(functionAnnotations, annotatedPythonFunction.getQualifiedName());
    }

    private void validateFunctionParameter(AnnotatedPythonParameter annotatedPythonParameter, String functionPath) {
        // TODO
    }

    private void validateConstructorParameter(AnnotatedPythonParameter annotatedPythonParameter, String functionPath) {
        // TODO
    }

    private void validateClass(AnnotatedPythonClass annotatedPythonClass) {
        annotatedPythonClass.getMethods().forEach(this::validateClassFunction);
        // TODO
    }

    private void validateModule(AnnotatedPythonModule annotatedPythonModule) {
        annotatedPythonModule.getClasses().forEach(this::validateClass);
        annotatedPythonModule.getFunctions().forEach(this::validateGlobalFunction);
        // TODO
    }

    private void validatePackage() {
        annotatedPythonPackage.getModules().forEach(this::validateModule);
        // TODO
    }

    private void validateAnnotationsValidOnTarget(
        EditorAnnotation[] editorAnnotations, AnnotatableDeclaration declaration, String qualifiedName
    ) {
        Set<String> validAnnotations = Set.of();
        switch (declaration) {
            case CLASS:
                validAnnotations = possibleClassAnnotations;
                break;
            case CLASS_FUNCTION:
                validAnnotations = possibleClassFunctionAnnotations;
                break;
            case CONSTRUCTOR_PARAMETER:
                validAnnotations = possibleConstructorParameterAnnotations;
                break;
            case GLOBAL_FUNCTION:
                validAnnotations = possibleGlobalFunctionAnnotations;
                break;
            case FUNCTION_PARAMETER:
                validAnnotations = possibleFunctionParameterAnnotations;
                break;
        }
        for (EditorAnnotation editorAnnotation: editorAnnotations) {
            String annotationName = getAnnotationName(editorAnnotation);
            if (annotationName != null && annotationName.endsWith(ANNOTATION_CLASS_SUFFIX)){
                annotationName = annotationName.substring(0, annotationName.length() - ANNOTATION_CLASS_SUFFIX.length());
            }
            if (!validAnnotations.contains(annotationName)) {
                validationErrors.add(new AnnotationTargetError(qualifiedName, annotationName, declaration));
            }
        }
    }

    private void validateAnnotationCombinations(EditorAnnotation[] editorAnnotations, String qualifiedName) {
        for (int i = 0; i < editorAnnotations.length; i++) {
            for (int j = i; j < editorAnnotations.length; j++) {
                validateAnnotationCombination(qualifiedName, editorAnnotations[i], editorAnnotations[j]);
            }
        }
    }

    private void validateAnnotationCombination(String qualifiedName, EditorAnnotation firstAnnotation, EditorAnnotation secondAnnotation) {
        String firstAnnotationName = getAnnotationName(firstAnnotation);
        String secondAnnotationName = getAnnotationName(secondAnnotation);
        if (!possibleCombinations.get(firstAnnotationName).contains(secondAnnotationName)) {
            validationErrors.add(
                new AnnotationCombinationError(qualifiedName,
                    firstAnnotationName,
                    secondAnnotationName)
            );
        }
    }

    private String getAnnotationName(EditorAnnotation editorAnnotation) {
        return editorAnnotation.getClass().getCanonicalName();
    }
}

File: api-editor/backend/src/main/kotlin/com/larsreimann/api_editor/validation/AnnotationValidator.kt
package com.larsreimann.api_editor.validation

import com.larsreimann.api_editor.model.AnnotationTarget
import com.larsreimann.api_editor.model.EditorAnnotation
import com.larsreimann.api_editor.model.GroupAnnotation
import com.larsreimann.api_editor.model.SerializablePythonClass
import com.larsreimann.api_editor.model.SerializablePythonFunction
import com.larsreimann.api_editor.model.SerializablePythonModule
import com.larsreimann.api_editor.model.SerializablePythonPackage
import com.larsreimann.api_editor.model.SerializablePythonParameter

class AnnotationValidator(private val annotatedPythonPackage: SerializablePythonPackage) {
    private val validationErrors = mutableListOf<AnnotationError>()
    private val groupAnnotationName = "Group"

    /**
     * Validates the classes annotated python package and returns validation errors.
     *
     * @return the validation errors found
     */
    fun validate(): List<AnnotationError> {
        validationErrors.clear()
        validatePackage()
        return validationErrors
    }

    private fun validatePackage() {
        annotatedPythonPackage.modules.forEach { validateModule(it) }
    }

    private fun validateModule(annotatedPythonModule: SerializablePythonModule) {
        annotatedPythonModule.classes.forEach { validateClass(it) }
        annotatedPythonModule.functions.forEach { validateGlobalFunction(it) }
    }

    private fun validateClass(annotatedPythonClass: SerializablePythonClass) {
        annotatedPythonClass.methods.forEach { validateMethod(it) }

        validateAnnotationsValidOnTarget(
            annotatedPythonClass.annotations,
            AnnotationTarget.CLASS,
            annotatedPythonClass.qualifiedName
        )
        validateAnnotationCombinations(
            annotatedPythonClass.annotations,
            annotatedPythonClass.qualifiedName
        )
    }

    private fun validateMethod(annotatedPythonFunction: SerializablePythonFunction) {
        val groupedParameterNames = getParameterNamesInGroups(annotatedPythonFunction.annotations)
        if (annotatedPythonFunction.isConstructor()) {
            annotatedPythonFunction.parameters.forEach {
                validateConstructorParameter(it, groupedParameterNames.contains(it.name))
            }
        } else {
            annotatedPythonFunction.parameters.forEach {
                validateFunctionParameter(it, groupedParameterNames.contains(it.name))
            }
        }

        validateAnnotationsValidOnTarget(
            annotatedPythonFunction.annotations,
            AnnotationTarget.METHOD,
            annotatedPythonFunction.qualifiedName
        )
        validateAnnotationCombinations(
            annotatedPythonFunction.annotations,
            annotatedPythonFunction.qualifiedName
        )
    }

    private fun validateGlobalFunction(annotatedPythonFunction: SerializablePythonFunction) {
        val groupedParameterNames = getParameterNamesInGroups(annotatedPythonFunction.annotations)
        annotatedPythonFunction.parameters.forEach {
            validateFunctionParameter(it, groupedParameterNames.contains(it.name))
        }

        validateAnnotationsValidOnTarget(
            annotatedPythonFunction.annotations,
            AnnotationTarget.GLOBAL_FUNCTION,
            annotatedPythonFunction.qualifiedName
        )
        validateAnnotationCombinations(
            annotatedPythonFunction.annotations,
            annotatedPythonFunction.qualifiedName
        )
    }

    private fun validateFunctionParameter(
        annotatedPythonParameter: SerializablePythonParameter,
        parameterInGroup: Boolean
    ) {
        validateAnnotationsValidOnTarget(
            annotatedPythonParameter.annotations,
            AnnotationTarget.FUNCTION_PARAMETER,
            annotatedPythonParameter.qualifiedName
        )
        validateAnnotationCombinations(
            annotatedPythonParameter.annotations,
            annotatedPythonParameter.qualifiedName
        )

        if (parameterInGroup) {
            validateGroupCombinations(
                annotatedPythonParameter.qualifiedName,
                annotatedPythonParameter.annotations
            )
        }
    }

    private fun validateConstructorParameter(
        annotatedPythonParameter: SerializablePythonParameter,
        parameterInGroup: Boolean
    ) {
        validateAnnotationsValidOnTarget(
            annotatedPythonParameter.annotations,
            AnnotationTarget.CONSTRUCTOR_PARAMETER,
            annotatedPythonParameter.qualifiedName
        )
        validateAnnotationCombinations(
            annotatedPythonParameter.annotations,
            annotatedPythonParameter.qualifiedName
        )

        if (parameterInGroup) {
            validateGroupCombinations(
                annotatedPythonParameter.qualifiedName,
                annotatedPythonParameter.annotations
            )
        }
    }

    private fun validateAnnotationsValidOnTarget(
        editorAnnotations: Iterable<EditorAnnotation>,
        target: AnnotationTarget,
        qualifiedName: String
    ) {
        for (editorAnnotation in editorAnnotations) {
            if (!editorAnnotation.isApplicableTo(target)) {
                validationErrors.add(AnnotationTargetError(qualifiedName, editorAnnotation.type, target))
            }
        }
    }

    private fun getParameterNamesInGroups(editorAnnotations: List<EditorAnnotation>): Set<String> {
        return editorAnnotations
            .filterIsInstance<GroupAnnotation>()
            .flatMap { it.parameters }
            .toSet()
    }

    private fun validateAnnotationCombinations(editorAnnotations: List<EditorAnnotation>, qualifiedName: String) {
        for (i in editorAnnotations.indices) {
            for (j in i + 1 until editorAnnotations.size) {
                validateAnnotationCombination(qualifiedName, editorAnnotations[i], editorAnnotations[j])
            }
        }
    }

    private fun validateAnnotationCombination(
        qualifiedName: String,
        firstAnnotation: EditorAnnotation,
        secondAnnotation: EditorAnnotation
    ) {
        val firstAnnotationName = firstAnnotation.type
        val secondAnnotationName = secondAnnotation.type
        if (firstAnnotationName !in possibleCombinations || secondAnnotationName !in possibleCombinations[firstAnnotationName]!!) {
            validationErrors.add(AnnotationCombinationError(qualifiedName, firstAnnotationName, secondAnnotationName))
        }
    }

    private fun validateGroupCombinations(qualifiedName: String, editorAnnotations: List<EditorAnnotation>) {
        editorAnnotations.forEach {
            val annotationName = it.type
            if (annotationName !in possibleCombinations[groupAnnotationName]!!) {
                validationErrors.add(GroupAnnotationCombinationError(qualifiedName, annotationName))
            }
        }
    }

    companion object {
        private var possibleCombinations = buildMap<String, Set<String>> {
            this["Attribute"] = mutableSetOf("Description", "Rename", "Todo")
            this["Boundary"] = mutableSetOf("Description", "Group", "Optional", "Rename", "Required", "Todo")
            this["CalledAfter"] = mutableSetOf("CalledAfter", "Description", "Group", "Move", "Pure", "Rename")
            this["Constant"] = mutableSetOf()
            this["Description"] = mutableSetOf(
                "Attribute",
                "Boundary",
                "CalledAfter",
                "Enum",
                "Group",
                "Move",
                "Optional",
                "Pure",
                "Rename",
                "Required", "Todo"
            )
            this["Enum"] = mutableSetOf("Description", "Group", "Rename", "Required", "Todo")
            this["Group"] =
                mutableSetOf(
                    "Boundary",
                    "CalledAfter",
                    "Description",
                    "Enum",
                    "Group",
                    "Move",
                    "Optional",
                    "Pure",
                    "Rename",
                    "Required",
                    "Todo"
                )
            this["Move"] = mutableSetOf("CalledAfter", "Description", "Group", "Pure", "Rename")
            this["Optional"] = mutableSetOf("Boundary", "Description", "Group", "Rename", "Todo")
            this["Pure"] = mutableSetOf("CalledAfter", "Description", "Group", "Move", "Rename")
            this["Remove"] = mutableSetOf()
            this["Rename"] = mutableSetOf(
                "Attribute",
                "Boundary",
                "CalledAfter",
                "Description",
                "Enum",
                "Group",
                "Move",
                "Optional",
                "Pure",
                "Required", "Todo"
            )
            this["Required"] = mutableSetOf("Boundary", "Description", "Enum", "Group", "Rename", "Todo")
            this["Todo"] = mutableSetOf(
                "Attribute",
                "Boundary",
                "Description",
                "Enum",
                "Group",
                "Optional",
                "Rename",
                "Required"
            )
        }
    }
}
